### PR TITLE
Minor: Add additional docstrings to Window function implementations

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

Related to #5781 

# Rationale for this change

As I begin working out how to expose User Defined Window Functions, I want to 
1. Make sure the documentation is up to date
2. Make sure I understand the code sufficiently

# What changes are included in this PR?
1. Add a bunch of doc comments to explain various window related functionality better
2. ~Make `PartitionEvaluator` public~ UPDATE: This PR just adds documentation, it doesn't change any APIs

# Are these changes tested?
Yes, with CI

# Are there any user-facing changes?
better rustdocs, hopefully

### CODE REVIEW

File: datafusion/physical-expr/src/window/built_in_window_function_expr.rs [hunk 1]
```diff
@@ -69,6 +73,8 @@ pub trait BuiltInWindowFunctionExpr: Send + Sync + std::fmt::Debug {
         false
     }
 
+    /// If returns true, [`Self::create_evaluator`] must implement
```
Comment (alamb) [inline datafusion/physical-expr/src/window/built_in_window_function_expr.rs:76]: I was a little unclear on why `use_window_frame` was part of `BuiltInWindowFunctionExpr` and not `PartitionEvaluator` but I haven't looked into it in more detail

File: datafusion/physical-expr/src/window/mod.rs [hunk 2]
```diff
@@ -32,6 +32,7 @@ mod window_frame_state;
 pub use aggregate::PlainAggregateWindowExpr;
 pub use built_in::BuiltInWindowExpr;
 pub use built_in_window_function_expr::BuiltInWindowFunctionExpr;
+pub use partition_evaluator::PartitionEvaluator;
```
Comment (alamb) [inline datafusion/physical-expr/src/window/mod.rs:35]: `PartitionEvaluator` now appears in public docstrings (and I propose to make it part of the user defined window function API -- see https://github.com/apache/arrow-datafusion/issues/5781#issuecomment-1581428228)

File: datafusion/physical-expr/src/window/partition_evaluator.rs [hunk 3]
```diff
@@ -25,24 +25,70 @@ use datafusion_common::{DataFusionError, ScalarValue};
...
+///  D  | 3
+/// ```
+///
+/// Will instantiate three `PartitionEvaluator`s, one each for the
+/// partitions defined by `col=A`, `col=B`, and `col=C`.
+///
+/// There are two types of `PartitionEvaluator`:
+///
+/// # Stateless `PartitionEvaluator`
+///
```
Comment (alamb) [inline datafusion/physical-expr/src/window/partition_evaluator.rs:52]: @mustafasrepo  / @ozankabak  if you have time to help me describe more clearly what Stateful and Stateless `PartitionEvaluator`s are , and specifically what is different between them, I would be most appreciative

File: datafusion/physical-expr/src/window/sliding_aggregate.rs [hunk 4]
```diff
@@ -36,12 +36,10 @@ use crate::{
...
-/// A window expr that takes the form of an aggregate function
-/// Aggregate Window Expressions that have the form
-/// `OVER({ROWS | RANGE| GROUPS} BETWEEN UNBOUNDED PRECEDING AND ...)`
-/// e.g cumulative window frames uses `PlainAggregateWindowExpr`. Where as Aggregate Window Expressions
-/// that have the form `OVER({ROWS | RANGE| GROUPS} BETWEEN M {PRECEDING| FOLLOWING} AND ...)`
-/// e.g sliding window frames uses `SlidingAggregateWindowExpr`.
+/// A window expr that takes the form of an aggregate function that
+/// can be incrementally computed over sliding windows.
+///
+/// See comments on [`WindowExpr`] for more details.
```
Comment (alamb) [inline datafusion/physical-expr/src/window/sliding_aggregate.rs:42]: Consolidated this description into `WindowExpr`

File: datafusion/physical-expr/src/window/partition_evaluator.rs (hunk 3 above)
Comment (alamb) [reply datafusion/physical-expr/src/window/partition_evaluator.rs:52]: Thank you @mustafasrepo  -- this is super helpful. I am incorporating this information into this PR

**Comment by alamb** [comment]
@mustafasrepo  / @ozankabak  I wonder if you have time to review this documentation and check if I got it right? This PR has no code changes in it

File: datafusion/physical-expr/src/window/built_in_window_function_expr.rs [hunk 5]
```diff
@@ -56,19 +62,36 @@ pub trait BuiltInWindowFunctionExpr: Send + Sync + std::fmt::Debug {
...
+    /// Create a [`PartitionEvaluator`] for evaluating the function on
+    /// a particular partition.
     fn create_evaluator(&self) -> Result<Box<dyn PartitionEvaluator>>;
 
-    /// Construct Reverse Expression that produces the same result
-    /// on a reversed window.  For example `lead(10)` --> `lag(10)`
+    /// Construct a new [`BuiltInWindowFunctionExpr`] that produces
+    /// the same result as this function on a window with reverse
+    /// order. The return value of this function is used by the
+    /// DataFusion optimizer to avoid resorting the data when
```
Comment (mustafasrepo) [inline datafusion/physical-expr/src/window/built_in_window_function_expr.rs:72]: Maybe we can use `re-sorting` instead of `resorting`.

**Review by mustafasrepo — APPROVED** [review]
Thanks @alamb for this PR. This PR is LGTM!.

File: datafusion/physical-expr/src/window/built_in_window_function_expr.rs (hunk 5 above)
Comment (alamb) [reply datafusion/physical-expr/src/window/built_in_window_function_expr.rs]: Done in d56a5b2e5

**Pull request merged**